numba/np/arraymath.py

```diff
@@ -3480,49 +3480,109 @@ def impl(arr, axis=None):
 
 
 @overload(np.delete)
-def np_delete(arr, obj):
+def np_delete(arr, obj, axis=None):
     # Implementation based on numpy
     # https://github.com/numpy/numpy/blob/af66e487a57bfd4850f4306e3b85d1dac3c70412/numpy/lib/function_base.py#L4065-L4267    # noqa: E501
 
     if not isinstance(arr, (types.Array, types.Sequence)):
         raise TypingError("arr must be either an Array or a Sequence")
 
+    if not isinstance(axis, (type(None), types.NoneType, types.Integer)):
+        raise TypingError("axis must be either a NoneType or an Integer")
+
     if isinstance(obj, (types.Array, types.Sequence, types.SliceType)):
         if isinstance(obj, (types.SliceType)):
             handler = np_delete_handler_isslice
+            obj_is_slice = True
         else:
             if not isinstance(obj.dtype, types.Integer):
                 raise TypingError('obj should be of Integer dtype')
             handler = np_delete_handler_isarray
+            obj_is_slice = False
+
+        if axis in (None, types.none):
+            def np_delete_impl(arr, obj, axis=None):
+                arr = np.ravel(np.asarray(arr))
+                N = arr.size
 
-        def np_delete_impl(arr, obj):
-            arr = np.ravel(np.asarray(arr))
-            N = arr.size
+                keep = np.ones(N, dtype=np.bool_)
+                obj = handler(obj)
+                keep[obj] = False
+                return arr[keep]
+        else:
+            ndim = arr.ndim
```

Comment:
This will fail when `arr` is a `types.Sequence`.

```python
import numpy as np
import numba

@numba.njit
def something(x, obj):
    return np.delete(x, obj, axis=1)

x = np.random.default_rng(1).random((3,3))

print(something.py_func(((1,2), (3,4)), slice(1,2))) # [[1][3]]
print(something(((1,2), (3,4)), slice(1,2))) # Fails
```
If we want to keep support for `types.Sequence`, the logic for following case needs to be updated such that doesn't require `ndim` to be calculated here.